Fix up fips integration

jyn514 opened the pull request:
The git history for this is kind of a mess, I can clean it up if you like.

This primarily does the following things:
- Build boringssl from source instead of requiring a pre-built version
- Adds an automated test to make sure that fips is always enabled when the feature is set
- Enforces in the build script that a supported version of clang is used to build boringssl when fips is enabled
- Adds a new boringssl-fips submodule, since `git submodule` only allows having one canonical commit checked-in per submodule

It also makes some minor cleanups:
- Fix tests that were broken with `--feature fips`
- Adds a CI job for testing with `--feature fips` enabled
- Fixes some hard-coded paths to match the source code layout in the older boringssl version
- Uses `X509_set_notBefore` unconditionally rather than only when fips is enabled

I removed the `BORING_BSSL_INCLUDE_PATH` support because it was a little tricky to get to work properly with FIPS and it should be possible to fix independently anyway.

cc @cjmakes

jyn514 committed:
[wip] try and fix fips tests

jyn514 committed:
fix typo in readme

jyn514 committed:
Use the same codepath with and without FIPS enabled

The version of boringssl that's FIPS-certified doesn't have `X509_set1_notAfter`.
The only difference between that and `X509_set_notAfter` is whether they're const-correct,
which doesn't seem worth having two different code-paths.

jyn514 committed:
[almost working] check out fips commit automatically

jyn514 committed:
Verify the version of the compiler used for building boringssl

NIST specifies that it needs to be 7.0.1; I originally tried building with clang 10 and it failed.
Theoretically this should check the versions of Go and Ninja too, but they haven't given me trouble in practice.

Example error:
```
   Compiling boring-sys v1.1.1 (/home/jnelson/work/boring/boring-sys)
error: failed to run custom build command for `boring-sys v1.1.1 (/home/jnelson/work/boring/boring-sys)`

Caused by:
  process didn't exit successfully: `/home/jnelson/work/boring/target/debug/build/boring-sys-31b8ce53031cfd83/build-script-build` (exit status: 101)
  --- stdout
  cargo:rerun-if-env-changed=BORING_BSSL_PATH

  --- stderr
  warning: missing clang-7, trying other compilers: Permission denied (os error 13)
  warning: FIPS requires clang version 7.0.1, skipping incompatible version "clang version 10.0.0-4ubuntu1 "
  thread 'main' panicked at 'unsupported clang version "cc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0": FIPS requires clang 7.0.1', boring-sys/build.rs:216:13
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

jyn514 committed:
Add Github actions workflow testing FIPS

jyn514 committed:
Update the README to remove BORING_BSSL_PATH mentions

behrat commented on README.md:
````diff
@@ -29,18 +29,9 @@ certified with [certificate
 ```bash
-$ cargo run --features fips-3678 --example fips_enabled
-boring::fips::enabled(): true
+$ cargo test --features fips-3678 fips::enabled
````
Since we now supply the boringssl-fips build with the library, I think it would make sense to suggest users to just use the `fips` feature (instead of a specific `fipx-XXXX` version), so that the boringssl-fips version is automatically updated if we ever update it in this library. My intention with the `fips-XXXX` was that we were explicit about what version of boringssl, as supplied by the user, we were linking against. Since I think this change also removed support for user supplying their own build via the `BORING_BSSL_INCLUDE_PATH` and `BORING_BSSL_LIB_PATH` variables, it might be least confusing to remove the `fips-3678` feature entirely. We can always add it back later if we actually do want to support multiple fips versions.

Also, I think you meant to use the test `is_enabled`. So, putting the two together, I think the docs should say:
```
cargo test --features fips fips::is_enabled 
```

jyn514 replied:
:+1: makes sense, I'll make both those changes

jyn514 committed:
Remove `fips-3678` feature

Now that build.rs checks out the appropriate commit automatically,
there's not a strong need to have the version number itself in the feature.
We can always add multiple features in the future if we support more than one FIPS-validated version.
